File: solutions/zig/queen-attack/1/queen_attack.zig
const std = @import("std");

pub const QueenError = error{
    InitializationFailure,
};

pub const Queen = struct {
    row: i8,
    col: i8,

    pub fn init(row: i8, col: i8) QueenError!Queen {
        if (row < 0 or row > 7 or col < 0 or col > 7) {
            return QueenError.InitializationFailure;
        }
        return Queen{ .row = row, .col = col };
    }

    pub fn canAttack(self: Queen, other: Queen) !bool {
        if (self.row == other.row and self.col == other.col) {
            return QueenError.InitializationFailure;
        }
        const row_diff = if (self.row > other.row) self.row - other.row else other.row - self.row;
        const col_diff = if (self.col > other.col) self.col - other.col else other.col - self.col;

        return self.row == other.row or self.col == other.col or row_diff == col_diff;
    }
};

File: solutions/typescript/queen-attack/1/queen-attack.ts
type Position = readonly [number, number];

type Positions = { white: Position; black: Position };

export class QueenAttack {
    public readonly black: Position;
    public readonly white: Position;

    constructor({ black = [0, 3], white = [7, 3] }: Partial<Positions> = {}) {
        if (this.isOffBoard(white)) {
            throw new Error('Queen must be placed on the board');
        }
        if (this.isOffBoard(black)) {
            throw new Error('Queen must be placed on the board');
        }
        if (this.areSamePosition(white, black)) {
            throw new Error('Queens cannot share the same space');
        }

        this.white = white;
        this.black = black;
    }

    private isOffBoard(position: Position): boolean {
        const [row, col] = position;
        return row < 0 || row > 7 || col < 0 || col > 7;
    }

    private areSamePosition(pos1: Position, pos2: Position): boolean {
        return pos1[0] === pos2[0] && pos1[1] === pos2[1];
    }

    toString(): string {
        const board = Array.from({ length: 8 }, () => Array(8).fill('_'));
        board[this.white[0]][this.white[1]] = 'W';
        board[this.black[0]][this.black[1]] = 'B';
        return board.map((row) => row.join(' ')).join('\n');
    }

    get canAttack(): boolean {
        if (
            this.white[0] === this.black[0] ||
            this.white[1] === this.black[1]
        ) {
            return true;
        }
        const rowDiff = Math.abs(this.white[0] - this.black[0]);
        const colDiff = Math.abs(this.white[1] - this.black[1]);
        return rowDiff === colDiff;
    }
}
